include/bluetooth/services/gadgets.h

```diff
@@ -0,0 +1,167 @@
+	BT_UUID_128_ENCODE(0xF04EB177, 0x3005, 0x43A7, 0xAC61, 0xA390DDF83076)
+
+/** @brief UUID of the RX Characteristic. **/
+#define GADGETS_UUID_GADGETS_RX_CHAR \
+	BT_UUID_128_ENCODE(0x2BEEA05B, 0x1879, 0x4BB4, 0x8A2F, 0x72641F82420B)
+
+#define BT_UUID_GADGETS_SERVICE BT_UUID_DECLARE_128(GADGETS_UUID_SERVICE)
+#define BT_UUID_GADGETS_RX BT_UUID_DECLARE_128(GADGETS_UUID_GADGETS_RX_CHAR)
+#define BT_UUID_GADGETS_TX BT_UUID_DECLARE_128(GADGETS_UUID_GADGETS_TX_CHAR)
+
+/** @brief Callback type for data received.
+ *         Returns false if type is unsupported.
+ */
+typedef bool (*gadgets_stream_cb_t)(
+				struct bt_conn *conn,
+				const uint8_t *const data,
+				uint16_t len,
+				bool more_data);
+
+/** @brief Callback type for data sent. */
+typedef void (*gadgets_sent_cb_t)(
+	struct bt_conn *conn, const void *buf, bool success);
+
+/** @brief Callback type for CCCD updated. */
+typedef void (*gadgets_cccd_update_cb_t)(bool enabled);
+
+/** @brief Pointers to the callback functions for service events. */
+struct bt_gatt_gadgets_cb {
+
+	/** Callback for incoming streams. **/
+	gadgets_stream_cb_t control_stream_cb;
+	gadgets_stream_cb_t alexa_stream_cb;
+
+	/** Callback for data sent. **/
+	gadgets_sent_cb_t sent_cb;
+
+	/** Callback for CCCD updated. **/
+	gadgets_cccd_update_cb_t cccd_cb;
+};
+
+/**@brief Initialize the service.
+ *
+ * @details This function registers the Gadgets BLE service with
+ *          two characteristics, TX and RX.
+ *
+ * @param[in] callbacks  Struct with function pointers to callbacks for service
+ *                       events.
+ *
+ * @retval 0 If initialization is successful.
+ *           Otherwise, a negative value is returned.
+ */
+int bt_gatt_gadgets_init(struct bt_gatt_gadgets_cb *callbacks);
+
+/**@brief Send data in stream format.
+ *
+ * @details This function encapsulates the data in the stream format
+ *          and transmits to the peer device.
+ *
+ * @param[in] conn Pointer to connection Object.
+ * @param[in] stream Stream type to use.
+ * @param[in] data Pointer to a data buffer. Buffer must be kept valid until
+ *          @ref gadgets_sent_cb_t is triggered.
+ * @param[in] size Size of the data in the buffer in bytes.
+ *
+ * @retval 0 If the data is sent.
+ *           Otherwise, a negative value is returned.
+ */
+int bt_gatt_gadgets_send_stream(
+	struct bt_conn *conn,
+	enum stream_id stream,
+	void *data,
+	size_t size);
+
+/**@brief Send data without stream formatting.
+ *
+ * @details No encapsulation or fragmentation is done of this data. Data buffer
+ *          does *not* need to be kept valid after this function returns.
+ *
+ * @param[in] conn Pointer to connection Object.
+ * @param[in] data Pointer to a data buffer.
+ * @param[in] size Size of the data in the buffer in bytes.
+ *
+ * @retval 0 If the data is sent.
+ *           Otherwise, a negative value is returned.
+ */
+int bt_gatt_gadgets_send(
+	struct bt_conn *conn,
+	void *data,
+	size_t size);
```

Comment:
The Gadgets service adds an additional layer of fragmentation on top of ATT, such that the maximum length is not limited by PDU, MTU, or ATT maximum sizes. The header uses uint16 as the length type, but I dont know if there are other limitations. I've not observed packet sizes above 600 bytes, but this depends on the configuration (capabilities, device name length etc.).
The format is described here (sub-page of developer docs referred to in gadgets.h): https://developer.amazon.com/en-US/docs/alexa/alexa-gadgets-toolkit/packet-ble.html#header

I can change this size parameter to uin16_t to more accurately reflect the underlying transport

Reply:
I got the stream and non-stream functions a bit confused here. The stream_send is limited to uint16_t datatype. The non-stream send is limited to MTU size (Notifications). The underlying bt_gatt_notify_cb will presumably return an error if the attempted transmission is larger than the current MTU.

Reply:
In that case, uint8_t will be enough for this one

Reply:
MTU can be larger than 255 bytes, I believe. For PDU it would be sufficient with uint8_t, but PDU size does not limit MTU size. (at least not per spec, perhaps things are different in practice)